cpp/src/groupby/sort/group_nth_element.cu

```diff
@@ -0,0 +1,113 @@
+/*
+ * Copyright (c) 2020, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <cudf/column/column_device_view.cuh>
+#include <cudf/column/column_factories.hpp>
+#include <cudf/column/column_view.hpp>
+#include <cudf/copying.hpp>
+#include <cudf/detail/aggregation/aggregation.hpp>
+#include <cudf/detail/gather.hpp>
+#include <cudf/detail/iterator.cuh>
+#include <cudf/types.hpp>
+
+#include <thrust/gather.h>
+
+namespace cudf {
+namespace experimental {
+namespace groupby {
+namespace detail {
+
+std::unique_ptr<column>
+group_nth_element(column_view const &values,
+                  column_view const &group_sizes,
+                  rmm::device_vector<size_type> const &group_labels,
+                  rmm::device_vector<size_type> const &group_offsets,
+                  size_type num_groups, size_type n,
+                  include_nulls _include_nulls,
+                  rmm::mr::device_memory_resource *mr, cudaStream_t stream) {
+
+  CUDF_EXPECTS(static_cast<size_t>(values.size()) == group_labels.size(),
+               "Size of values column should be same as that of group labels");
+
+  if (num_groups == 0) {
+    return cudf::experimental::empty_like(values);
+  }
+
+  auto output = make_numeric_column(
+      cudf::data_type{cudf::experimental::type_to_id<size_type>()}, num_groups,
+      cudf::UNALLOCATED, stream, mr); // UNINITIALIZED
```

Comment:
`mr` from the caller should not be used for allocations that are not to be returned to caller.
```suggestion
      cudf::UNALLOCATED, stream); // UNINITIALIZED
```